cargo-miri/bin.rs

```diff
@@ -568,24 +602,36 @@ fn phase_cargo_rustc(args: env::Args) {
         _ => {},
     }
 
-    if !print && target_crate && is_runnable_crate() {
-        // This is the binary or test crate that we want to interpret under Miri.
-        // But we cannot run it here, as cargo invoked us as a compiler -- our stdin and stdout are not
-        // like we want them.
-        // Instead of compiling, we write JSON into the output file with all the relevant command-line flags
-        // and environment variables; this is used when cargo calls us again in the CARGO_TARGET_RUNNER phase.
-        let info = CrateRunInfo::collect(args);
-        let filename = out_filename("", "");
-        if verbose {
-            eprintln!("[cargo-miri rustc] writing run info to `{}`", filename.display());
+    if !print && is_runnable_crate() {
+        macro_rules! store {
+            ($info:expr) => {
+                let info = $info;
+                let filename = out_filename("", "");
+                if verbose {
+                    eprintln!("[cargo-miri rustc] writing run info to `{}`", filename.display());
+                }
+                info.store(&filename);
+                // For Windows, do the same thing again with `.exe` appended to the filename.
+                // (Need to do this here as cargo moves that "binary" to a different place before
+                // running it.)
+                info.store(&out_filename("", ".exe"));
+                return;
+            };
+        }
```

Comment:
Does this have to be a macro? I'd much prefer something closure-based, with proper types.

Reply:
I changed to a closure. I used a macro to avoid repeating `return`.